dlfcn_android.go

```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2024 The Ebitengine Authors
+
+package purego
+
+import "github.com/ebitengine/purego/internal/cgo"
+
+var (
```

Comment:
If we use var here, we should also change other platforms to var for consistency. What do you think?

Reply:
I prefer them being constants since it makes it clear they aren't supposed to be modified. Android is just "unique"

Reply:
So this code would no longer be portable, though we would not do that

```go
var _ [purego.RTLD_GLOBAL]int
```

Reply:
Sure but why would anyone do that? Making it var breaks compatibility for all the people who do currently do that instead of just on android which is likely less common anyways

Reply:
>  Making it var breaks compatibility for all the people who do currently do that instead of just on android which is likely less common anyways

The version is still 0.x so it is still allowed to break compatibility.

Another example would be:

```go
_ = math.Abs(purego.RTLD_GLOBAL) // This works only when RTLD_GLOBAL is an untyped constant.
```

```go
type T int

func main() {
    var _ T = purego.RTLD_GLOBAL // This works only when RTLD_GLOBAL is an untyped constant.
}
```

At least, if we keep constants, we should add types (int) instead of untyped constants, so that the both cases would be portable and no longer be compiled.

Reply:
Agreed they shouldn't be untype constants but constant int

Reply:
Hmmm... This seems to break more things...

```
./dlfcn_test.go:23:28: cannot use purego.RTLD_DEFAULT (constant -2 of type int) as uintptr value in argument to purego.Dlsym
./dlfcn_test.go:91:31: cannot use purego.RTLD_DEFAULT (constant -2 of type int) as uintptr value in argument to purego.Dlsym
./dlfcn_test.go:94:40: cannot use purego.RTLD_DEFAULT (constant -2 of type int) as uintptr value in argument to purego.SyscallN
```

It's because Dlopen takes an int but Dlysm and SyscallN take uintptr. Not sure which to change since it currently matches the C header (int -> int, void* -> uintptr)

Reply:
What about uintptr constants (and variables)? Wouldn't -2 work?

Reply:
https://github.com/ebitengine/purego/blob/main/dlfcn_freebsd.go#L8

Yeah I feel like we should define uintptr constants (and variables)

Reply:
And we would have to change `mode int` to `mode uintptr`, maybe...?

https://pkg.go.dev/github.com/ebitengine/purego#Dlopen

Reply:
What do you mean constants and variables? Only RTLD_DEFAULT needs to uintptr. Wasn't it the case that was the only one on android that wasn't const so perhaps?

```
var
RTLD_DEFAULT = cgo.RTLD_DEFAULT
)
const(
	RTLD_LAZY    = cgo.RTLD_LAZY
	RTLD_NOW     = cgo.RTLD_NOW
	RTLD_LOCAL   = cgo.RTLD_LOCAL
	RTLD_GLOBAL  = cgo.RTLD_GLOBAL
)

```

Reply:
Hm? Can we use consts for cgo values?

Reply:
Yes as long as it's value is constant which it is for all of those except RTLD_DEFAULT

Reply:
By the way, for RTLD_*, what about defining values by ourselves form [the header](https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/include/dlfcn.h)? I mean

```go
//go:build android && (386 || arm)

const (
	RTLD_DEFAULT = 0xffffffff
	RTLD_LAZY    = 0x00000001
	RTLD_NOW     = 0x00000000
	RTLD_LOCAL   = 0x00000000
	RTLD_GLOBAL  = 0x00000002
)
```

Reply:
> Yes as long as it's value is constant which it is for all of those except RTLD_DEFAULT [at least on linux]

Interesting. So only `cgo.RTLD_DEFAULT` cannot be a constant, right?

Reply:
Yes because it's being casted in the header to a `void*` although its value doesn't change

Reply:
> By the way, for RTLD_*, what about defining values by ourselves form the header? I mean

I would, but I thought we determined that that doesn't always work on android. Am I mistaken?

Reply:
> I thought we determined that that doesn't always work on android. Am I mistaken?

My understanding is that `purego.Dlopen` succeeded with the constants for 32bit Android, but a mysterious crash happens after that. The crash should be fixed by using C's dl functions directly by this PR.

Reply:
Let's try with the constants hardcoded in dlfcn_android.go

Reply:
Thanks!

Reply:
Not sure if my approach is the best since now there are 3 extra files but the easiest thing I could come up with so we can test it

Reply:
We can reduce one file by merging `dlfcn_linux.go` and `dlfcn_android_64bit.go`, but I am not sure which is better.